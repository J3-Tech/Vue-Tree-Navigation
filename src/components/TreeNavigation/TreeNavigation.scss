.TreeNavigation {
  display: inline-block;
  padding: 0;
  margin:  0;

  ul {
    padding:         0;
    margin:          0;
    list-style-type: none;
  }

  li {
    padding-left: 20px;
  }
}
