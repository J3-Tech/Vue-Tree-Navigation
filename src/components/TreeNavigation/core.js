import NavigationLevel from '../NavigationLevel/NavigationLevel.vue';
import NavigationItem from '../NavigationItem/NavigationItem.vue';

import {
  PATH_TYPE_NONE,
  PATH_TYPE_ELEMENT,
  PATH_TYPE_ROUTE,
  PATH_TYPE_EXTERNAL,
} from '../../config';

import {
  sanitizeElement,
  sanitizeRoute,
  removeElementFromPath,
  getRelativeUrl,
} from '../utils';

/**
 * Recursive function.
 * One call generates one level of the tree.
 */
export const generateLevel = (
  createElement,
  items,
  level,
  defaultOpenLevel
) => {
  const children = [];

  items.forEach(item => {
    if (item.hasOwnProperty('children')) {
      const navLevel = createElement(
        NavigationLevel,
        {
          props: {
            level,
            open: renderLevelAsOpen(item, level, defaultOpenLevel),
            parentItem: item,
          },
        },
        [
          ...generateLevel(
            createElement,
            item.children,
            level + 1,
            defaultOpenLevel
          ),
        ]
      );

      children.push(createElement('li', [navLevel]));
    } else {
      const navItem = createElement(NavigationItem, {
        props: {
          item,
        },
      });

      children.push(createElement('li', [navItem]));
    }
  });

  return children;
};

/**
 * Level should be opened in following cases
 * - level is less than or equal to default open level
 * - it contains child which URL is a part of an active URL
 */
export const renderLevelAsOpen = (parentItem, level, defaultOpenLevel) => {
  if (defaultOpenLevel >= level) {
    return true;
  }

  const currentUrl = getRelativeUrl(
    window.location.href,
    window.location.origin
  );

  for (let i = 0; i < parentItem.children.length; i++) {
    let child = parentItem.children[i];

    if (child.meta.path !== undefined) {
      let childUrl = child.meta.path;

      if (!childUrl.startsWith('/#')) {
        childUrl = removeElementFromPath(childUrl);
      }

      if (currentUrl.startsWith(childUrl)) {
        return true;
      }
    }
  }

  return false;
};

/**
 * Recursive function.
 * Insert metadata containing the navigation path and its type to each item.
 **/
export const insertMetadataToItems = (items, parent) => {
  items.forEach(item => {
    item.meta = getItemMetadata(item, parent);

    if (item.hasOwnProperty('children')) {
      item.children = insertMetadataToItems(item.children, item);
    }
  });

  return items;
};

/**
 * Return item metadata object: { path: ..., pathType: ... }
 */
export const getItemMetadata = (item, parent) => {
  const element = sanitizeElement(item.element);
  const route = sanitizeRoute(item.route);
  const external = item.external;

  // item is its own parent
  if (parent === undefined) {
    if (
      element === undefined &&
      route === undefined &&
      external === undefined
    ) {
      return {
        path: undefined,
        pathType: PATH_TYPE_NONE,
      };
    }

    if (external !== undefined) {
      return {
        path: external,
        pathType: PATH_TYPE_EXTERNAL,
      };
    }

    if (route !== undefined) {
      return {
        path: route,
        pathType: PATH_TYPE_ROUTE,
      };
    }

    if (element !== undefined) {
      return {
        path: element,
        pathType: PATH_TYPE_ELEMENT,
      };
    }
  }

  // route -> external
  if (parent.meta.pathType === PATH_TYPE_ROUTE && external !== undefined) {
    return {
      path: external,
      pathType: PATH_TYPE_EXTERNAL,
    };
  }

  // route -> route
  if (parent.meta.pathType === PATH_TYPE_ROUTE && route !== undefined) {
    const parentPath = removeElementFromPath(parent.meta.path);

    return {
      path: parentPath + route,
      pathType: PATH_TYPE_ROUTE,
    };
  }

  // route -> element
  if (parent.meta.pathType === PATH_TYPE_ROUTE && element !== undefined) {
    const parentPath = removeElementFromPath(parent.meta.path);

    return {
      path: parentPath + element,
      pathType: PATH_TYPE_ROUTE,
    };
  }

  // route -> label
  if (
    parent.meta.pathType === PATH_TYPE_ROUTE &&
    element === undefined &&
    route === undefined &&
    external === undefined
  ) {
    const parentPath = removeElementFromPath(parent.meta.path);

    return {
      path: parentPath,
      pathType: PATH_TYPE_ROUTE,
    };
  }

  // element -> external
  if (parent.meta.pathType === PATH_TYPE_ELEMENT && external !== undefined) {
    return {
      path: external,
      pathType: PATH_TYPE_EXTERNAL,
    };
  }

  // element -> route
  if (parent.meta.pathType === PATH_TYPE_ELEMENT && route !== undefined) {
    return {
      path: route,
      pathType: PATH_TYPE_ROUTE,
    };
  }

  // element -> element
  if (parent.meta.pathType === PATH_TYPE_ELEMENT && element !== undefined) {
    return {
      path: element,
      pathType: PATH_TYPE_ELEMENT,
    };
  }

  // element -> label
  if (
    parent.meta.pathType === PATH_TYPE_ELEMENT &&
    element === undefined &&
    route === undefined &&
    external === undefined
  ) {
    return {
      path: undefined,
      pathType: PATH_TYPE_NONE,
    };
  }

  // label -> external
  if (parent.meta.pathType === PATH_TYPE_NONE && external !== undefined) {
    return {
      path: external,
      pathType: PATH_TYPE_EXTERNAL,
    };
  }

  // label -> route
  if (parent.meta.pathType === PATH_TYPE_NONE && route !== undefined) {
    return {
      path: route,
      pathType: PATH_TYPE_ROUTE,
    };
  }

  // label -> element
  if (parent.meta.pathType === PATH_TYPE_NONE && element !== undefined) {
    return {
      path: '/' + element,
      pathType: PATH_TYPE_ELEMENT,
    };
  }

  // label -> label
  if (
    parent.meta.pathType === PATH_TYPE_NONE &&
    element === undefined &&
    route === undefined
  ) {
    return {
      path: undefined,
      pathType: PATH_TYPE_NONE,
    };
  }

  // external -> external
  if (parent.meta.pathType === PATH_TYPE_EXTERNAL && external !== undefined) {
    return {
      path: external,
      pathType: PATH_TYPE_EXTERNAL,
    };
  }

  // external -> route
  if (parent.meta.pathType === PATH_TYPE_EXTERNAL && route !== undefined) {
    return {
      path: route,
      pathType: PATH_TYPE_ROUTE,
    };
  }

  // external -> element
  if (parent.meta.pathType === PATH_TYPE_EXTERNAL && element !== undefined) {
    return {
      path: element,
      pathType: PATH_TYPE_ELEMENT,
    };
  }

  // external -> label
  if (
    parent.meta.pathType === PATH_TYPE_EXTERNAL &&
    element === undefined &&
    route === undefined &&
    external === undefined
  ) {
    return {
      path: undefined,
      pathType: PATH_TYPE_NONE,
    };
  }
};
