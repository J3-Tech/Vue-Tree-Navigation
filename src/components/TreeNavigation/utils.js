import NavigationList from '../NavigationList/NavigationList.vue';

/**
 * Get navigation item - a hyperlink, a router link, or a simple value.
 * Return only `name` in case both `to` and `href` are not defined.
 * Return <router-link to=`location`>`name`</router-link> if `to` is defined.
 * Return <a href=`location`>`name`</a> when `href` is defined.
 */
function getNavItem(createElement, { name, to, href }, location='') {

  if (to === undefined && href === undefined) {
    return name;
  }

  let element;
  let attrs;
  let props;

  if (href !== undefined) {
    element = 'a';
    attrs = {
      href: location,
    };
    props = {};
  }

  // user shouldn't define both `to` and `href` but in case he does,
  // `href` will be overridden by `to` which has a higher priority
  if (to !== undefined) {
    element = 'router-link';
    attrs = {};
    props = {
      to: location,
    };
  }

  return createElement(element, {
    attrs,
    props,
  }, [
    name,
  ]);
}

/**
 * Recursive function.
 * One call generates one level of the tree.
 */
function generateLevel(createElement, items, level, defaultOpenLevel) {
  const children = [];

  items.forEach(item => {
    const navItem = getNavItem(createElement, item);

    if (item.hasOwnProperty('children')) {
      children.push(
        createElement(NavigationList, {
          props: {
            level,
            defaultOpenLevel,
          },
        }, [
          navItem,
          ...generateLevel(createElement,
                           item.children,
                           level + 1,
                           defaultOpenLevel),
        ])
      );
    } else {
      children.push(createElement('li', [
        navItem,
      ]));
    }
  });

  return children;
}

export default {
  generateLevel,
};
