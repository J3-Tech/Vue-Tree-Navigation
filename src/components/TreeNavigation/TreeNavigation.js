import utils from './utils';

require('./assets/_base.scss');

const TreeNavigation = {
  name: 'vue-tree-navigation',

  data() {
    return {
      classes: ['tree-navigation'],
    };
  },

  props: {
    items: {
      type: Array,
      required: true,
    },
    defaultOpenLevel: {
      type: Number,
      default: 0,
    },
  },

  computed: {
    itemsWithMetadata() {
      const self = this;

      const items = JSON.parse(JSON.stringify(self.items));
      return utils.insertMetadataToItems(items);
    },
  },

  render(createElement) {
    const self = this;

    const level = 1;
    const tree = createElement('ul',
      utils.generateLevel(createElement,
                          self.itemsWithMetadata,
                          level,
                          self.defaultOpenLevel));

    const wrapper = createElement('div', {
      'class': self.classes,
    }, [
      tree,
    ]);

    return wrapper;
  },
};

export default TreeNavigation;
