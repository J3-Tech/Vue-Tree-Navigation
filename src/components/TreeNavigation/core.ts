import NavigationLevel from '../NavigationLevel/NavigationLevel.vue';
import NavigationItem from '../NavigationItem/NavigationItem.vue';

import { sanitizeElement, sanitizePath } from '../utils';

/**
 * Recursive function.
 * One call generates one level of the tree.
 */
export const generateLevel = (
  createElement,
  items,
  level,
  defaultOpenLevel
) => {
  const children = [];

  items.forEach(item => {
    if (item.hasOwnProperty('children')) {
      const navLevel = createElement(
        NavigationLevel,
        {
          props: {
            parentItem: item,
            level,
            defaultOpenLevel,
          },
        },
        [
          ...generateLevel(
            createElement,
            item.children,
            level + 1,
            defaultOpenLevel
          ),
        ]
      );

      children.push(createElement('li', [navLevel]));
    } else {
      const navItem = createElement(NavigationItem, {
        props: {
          item,
        },
      });

      children.push(createElement('li', [navItem]));
    }
  });

  return children;
};

/**
 * Recursive function.
 * Insert metadata containing the navigation path and its type to each item.
 **/
export const insertMetadataToNavItems = (items, parent) => {
  items.forEach(item => {
    item.meta = getItemMetadata(item, parent);

    if (item.hasOwnProperty('children')) {
      item.children = insertMetadataToNavItems(item.children, item);
    }
  });

  return items;
};

/**
 * Return item metadata object: { path: ..., target: ... }
 */
export const getItemMetadata = (item, parent) => {
  const element = sanitizeElement(item.element);
  const path = sanitizePath(item.path);
  const external = item.external;

  // item is its own parent
  if (parent === undefined) {
    if (element === undefined && path === undefined && external === undefined) {
      return {
        path: '',
        target: '',
      };
    }

    if (external !== undefined) {
      return {
        path: '',
        target: external,
      };
    }

    if (path !== undefined) {
      return {
        path: path,
        target: path,
      };
    }

    if (element !== undefined) {
      return {
        path: '',
        target: '/' + element,
      };
    }
  }

  const parentPath = sanitizePath(parent.meta.path);

  if (external !== undefined) {
    return {
      path: parentPath,
      target: external,
    };
  }

  if (path !== undefined) {
    return {
      path: parentPath + path,
      target: parentPath + path,
    };
  }

  if (element !== undefined) {
    return {
      path: parentPath,
      target: sanitizePath(parentPath + element),
    };
  }

  return {
    path: parentPath,
    target: '',
  };
};
