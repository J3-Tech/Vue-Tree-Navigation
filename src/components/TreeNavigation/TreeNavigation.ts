import { insertMetadataToNavItems, generateLevel } from './core';

import './TreeNavigation.scss';

const TreeNavigation = {
  props: {
    items: {
      type: Array,
      required: false,
      default: () => [],
    },
    defaultOpenLevel: {
      type: Number,
      default: 0,
    },
  },
  computed: {
    navItems() {
      if (this.items && this.items.length) {
        return this.items;
      }

      if (
        this.$router &&
        this.$router.options &&
        this.$router.options.routes &&
        this.$router.options.routes.length
      ) {
        return this.$router.options.routes;
      }

      console.warn(
        "[VueTreeNavigation]: Haven't you forget to provide items or define vue-router routes?"
      );
      return [];
    },
    navItemsWithMetadata() {
      const navItems = JSON.parse(JSON.stringify(this.navItems));
      return insertMetadataToNavItems(navItems);
    },
  },

  render(createElement) {
    const self = this;

    const level = 1;
    const tree = createElement(
      'ul',
      generateLevel(
        createElement,
        self.navItemsWithMetadata,
        level,
        self.defaultOpenLevel
      )
    );

    const level0 = createElement(
      'div',
      {
        class: ['NavigationLevel', 'NavigationLevel--level-0'],
      },
      [tree]
    );

    const treeNavigation = createElement(
      'div',
      {
        class: 'TreeNavigation',
      },
      [level0]
    );

    return treeNavigation;
  },
};

export default TreeNavigation;
