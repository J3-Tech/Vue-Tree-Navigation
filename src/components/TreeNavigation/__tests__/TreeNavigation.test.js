import { createLocalVue, mount } from 'vue-test-utils';
import VueRouter from 'vue-router';

import TreeNavigation from '../TreeNavigation';

const items = [
  { name: 'Home', to: 'home' },
  { name: 'Products', to: 'products', children: [
    { name: 'Running shoes', to: 'shoes', children: [
      { name: 'Race', href: '#race' },
      { name: 'Road', href: '#road' },
      { name: 'Trail', href: '#trail' },
    ]},
    { name: 'Running clothing', to: 'clothing', children: [
      { name: 'Jackets', to: 'jackets' },
      { name: 'Tops', to: 'tops', children: [
        { name: 'Long Sleeve', href: '#long-sleeve', children: [
          { name: 'For summer', href: '#summer' },
          { name: 'For winter', href: '#winter' },
        ]},
        { name: 'Short Sleeve', href: '#short-sleeve' },
        { name: 'Sleeveless', href: '#sleeveless' },
      ]},
    ]},
  ]},
  { name: 'About', children: [
    { name: 'Company' },
    { name: 'Contact', href: '#contact' },
    { name: 'Blog', href: '#blog' },
  ]},
];

const routes = [
  {
    path: '/home',
  },
  {
    path: '/products',
  },
  {
    path: '/products/shoes',
  },
  {
    path: '/products/clothing',
    children: [
      {
        path: 'jackets',
      },
      {
        path: 'tops',
      },
    ],
  },
];

describe('TreeNavigation', () => {
  let localVue;
  let router;

  beforeAll(() => {
    localVue = createLocalVue();
    localVue.use(VueRouter);
    router = new VueRouter({
      routes,
    });
  });

  it('renders', () => {
    const wrapper = mount(TreeNavigation, {
      localVue,
      router,
      propsData: {
        items,
      },
    });

    // link targets are dummy # right now
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('item', () => {
    context('with `to` and `href` fields undefined', () => {
      it('is rendered as a simple value', () => {
        const wrapper = mount(TreeNavigation, {
          propsData: {
            items: [{
              name: 'Item',
            }],
          },
        });

        expect(wrapper.find('li').html()).toBe('<li>Item</li>');
      });
    });

    context('with `href` field defined', () => {
      it('is rendered as a hyperlink', () => {
        const wrapper = mount(TreeNavigation, {
          propsData: {
            items: [{
              name: 'Item',
              href: '#contact',
            }],
          },
        });

        expect(wrapper.find('li').contains('a')).toBe(true);
      });

      it('hyperlink is not assigned `router-link` class', () => {
        const wrapper = mount(TreeNavigation, {
          propsData: {
            items: [{
              name: 'Item',
              href: '#contact',
            }],
          },
        });

        expect(wrapper.find('a').classes()).not.toContain('router-link');
      });
    });

    context('with `to` field defined', () => {
      it('is rendered as a router link', () => {
        const wrapper = mount(TreeNavigation, {
          propsData: {
            items: [{
              name: 'Item',
              to: 'home',
            }],
          },
        });

        expect(wrapper.find('li').contains('router-link')).toBe(true);
      });

      it('router link is assigned `router-link` class', () => {
        const wrapper = mount(TreeNavigation, {
          propsData: {
            items: [{
              name: 'Item',
              to: 'home',
            }],
          },
        });

        expect(wrapper.find('router-link').classes()).toContain('router-link');
      });
    });
  });

  describe('lists', () => {
    context('with level deeper than default open level', () => {
      it('are assigned `closed` class', () => {
        const wrapper = mount(TreeNavigation, {
          propsData: {
            items,
            defaultOpenLevel: 2,
          },
        });

        // level 3 list
        expect(wrapper.find('div > ul > ul > ul > ul').classes()).toContain('closed');

        // level 4 list
        expect(wrapper.find('div > ul > ul > ul > ul > ul').classes()).toContain('closed');
      });
    });

    context('with level less than or equal to default open level', () => {
      it('are not assigned `closed` class', () => {
        const wrapper = mount(TreeNavigation, {
          propsData: {
            items,
            defaultOpenLevel: 2,
          },
        });

        // level 0 list
        expect(wrapper.find('div > ul').classes()).not.toContain('closed');

        // level 1 list
        expect(wrapper.find('div > ul > ul').classes()).not.toContain('closed');

        // level 2 list
        expect(wrapper.find('div > ul > ul > ul').classes()).not.toContain('closed');
      });
    });
  });
});
