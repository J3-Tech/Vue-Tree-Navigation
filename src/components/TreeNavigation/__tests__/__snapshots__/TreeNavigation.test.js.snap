// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeNavigation renders 1`] = `
<div class="tree-navigation">
    <ul>
        <li>
            <a href="#/#" class="router-link">Home</a>
        </li>
        <ul class="navigation-list closed"><span class="navigation-toggle closed"><span class="navigation-toggle-icon"></span></span>
            <a href="#/#" class="router-link">Products</a>
            <ul class="navigation-list closed"><span class="navigation-toggle closed"><span class="navigation-toggle-icon"></span></span>
                <a href="#/#" class="router-link">Running shoes</a>
                <li>
                    <a href="#" class="">Race</a>
                </li>
                <li>
                    <a href="#" class="">Road</a>
                </li>
                <li>
                    <a href="#" class="">Trail</a>
                </li>
            </ul>
            <ul class="navigation-list closed"><span class="navigation-toggle closed"><span class="navigation-toggle-icon"></span></span>
                <a href="#/#" class="router-link">Running clothing</a>
                <li>
                    <a href="#/#" class="router-link">Jackets</a>
                </li>
                <ul class="navigation-list closed"><span class="navigation-toggle closed"><span class="navigation-toggle-icon"></span></span>
                    <a href="#/#" class="router-link">Tops</a>
                    <ul class="navigation-list closed"><span class="navigation-toggle closed"><span class="navigation-toggle-icon"></span></span>
                        <a href="#" class="">Long Sleeve</a>
                        <li>
                            <a href="#" class="">For summer</a>
                        </li>
                        <li>
                            <a href="#" class="">For winter</a>
                        </li>
                    </ul>
                    <li>
                        <a href="#" class="">Short Sleeve</a>
                    </li>
                    <li>
                        <a href="#" class="">Sleeveless</a>
                    </li>
                </ul>
            </ul>
        </ul>
        <ul class="navigation-list closed"><span class="navigation-toggle closed"><span class="navigation-toggle-icon"></span></span> About
            <li>Company</li>
            <li>
                <a href="#" class="">Contact</a>
            </li>
            <li>
                <a href="#" class="">Blog</a>
            </li>
        </ul>
    </ul>
</div>
`;
